elastic/beats
# Update k8s client and release k8s lock gracefully

## What does this PR do?
Updates k8s go client to utilise a [fix](https://github.com/kubernetes/kubernetes/pull/80954) so as to gracefully release leader lock.

## Why is it important?
To release the leader lease gracefully and make it instantly available for other candidates.


## Related issues

- Closes https://github.com/elastic/beats/issues/21952
- Relates https://github.com/elastic/beats/pull/21896


## How to test this PR locally

1. Deploy a k8s cluster using kind: `kind create cluster --config kind-mutly.yaml  `
```
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
- role: worker
- role: worker
```
2. Deploy metricbeat with the latest change included using 7.11.0-SNAPSHOT

[metricbeat-kubernetes-leader.7.11.yaml.txt](https://github.com/elastic/beats/files/5642178/metricbeat-kubernetes-leader.7.11.yaml.txt)

3. Follow the logs of the 2 daemonset pods and kill the one holding the lease. Verify that no [errors](https://github.com/elastic/beats/pull/21896#issue-504753785) occur and that the lease was gained by the other Pod.

4. Verify that basic autodiscovery works along with metadata:
[nats.yaml.txt](https://github.com/elastic/beats/files/5642307/nats.yaml.txt)
